src/cascadia/TerminalApp/Tab.cpp

```diff
@@ -513,15 +519,164 @@ namespace winrt::TerminalApp::implementation
+        }
+        else
+        {
+            _ConstructTabRenameBox(tabText);
+        }
+    }
+
+    // Method Description:
+    // - Create a new TextBox to use as the control for renaming the tab text.
+    //   If the text box is already created, then this will do nothing, and
+    //   leave the current box unmodified.
+    // Arguments:
+    // - tabText: This should be the text to initialize the rename text box with.
+    // Return Value:
+    // - <none>
+    void Tab::_ConstructTabRenameBox(const winrt::hstring& tabText)
+    {
+        if (_tabViewItem.Header().try_as<Controls::TextBox>())
+        {
+            return;
+        }
+
+        Controls::TextBox tabTextBox;
+        tabTextBox.Text(tabText);
+
+        // The TextBox has a MinHeight already set by default, which is
+        // larger than we want. Get rid of it.
+        tabTextBox.MinHeight(0);
+        // Also get rid of the internal padding on the text box, between the
+        // border and the text content, on the top and bottom. This will
+        // help the box fit within the bounds of the tab.
+        Thickness internalPadding = ThicknessHelper::FromLengths(4, 0, 4, 0);
+        tabTextBox.Padding(internalPadding);
+
+        // Make the margin (0, -8, 0, -8), to counteract the padding that
+        // the TabViewItem has.
+        //
+        // This is maybe a bit fragile, as the actual value might not be exactly
+        // (0, 8, 0, 8), but using TabViewItemHeaderPadding to look up the real
+        // value at runtime didn't work. So this is good enough for now.
+        Thickness negativeMargins = ThicknessHelper::FromLengths(0, -8, 0, -8);
+        tabTextBox.Margin(negativeMargins);
+
+        // Set up some event handlers on the text box. We need three of them:
+        // * A LostFocus event, so when the TextBox loses focus, we'll
+        //   remove it and return to just the text on the tab.
+        // * A KeyUp event, to be able to submit the tab text on Enter or
+        //   dismiss the text box on Escape
+        // * A LayoutUpdated event, so that we can auto-focus the text box
+        //   when it's added to the tree.
+        auto weakThis{ get_weak() };
+
+        // When the text box loses focus, update the tab title of our tab.
+        // - If there are any contents in the box, we'll use that value as
+        //   the new "runtime text", which will override any text set by the
+        //   application.
+        // - If the text box is empty, we'll reset the "runtime text", and
+        //   return to using the active terminal's title.
+        tabTextBox.LostFocus([weakThis](const IInspectable& sender, auto&&) {
+            auto tab{ weakThis.get() };
+            auto textBox{ sender.try_as<Controls::TextBox>() };
+            if (tab && textBox)
+            {
+                tab->_runtimeTabText = textBox.Text();
+                tab->_inRename = false;
+                tab->_UpdateTitle();
+            }
+        });

```

Comment:
I think this might be the wrong behavior, but I guess I need to build the branch to come to a decision here. In my mind, if I start renaming a tab, then I use my mouse to click somewhere else, I would expect the operation to be cancelled, not completed.

Reply:
Actually IMHO the expected behavior should be like implemented: Blurring (with the exception of by pressing <kbd>ESC</kbd>) should _commit_ the change.

Reply:
I agree. Loss of focus is "commit" or "validate" in almost every application. <kbd>Esc</kbd> is cancel.

Reply:
I'm gonna resolve this thread for now, since this does seem to be a pretty standard UX pattern